### PR TITLE
Add information to header

### PR DESCRIPTION
**Pull request by benjaminjb**
**Checklist:**
 - [ ] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [ ] Have you tested your changes on all related environments with successful results, as applicable?
   - [ ] Have you added automated tests?

**Type of Changes:**
 - [ ] New feature
 - [ ] Bug fix
 - [ ] Documentation
 - [ ] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Other Information**:
Issues: [PGO-1610, PGO-1616, PGO-1618]

### CODE REVIEW

File: internal/upgradecheck/header_test.go [hunk 1]
```diff
@@ -500,12 +514,35 @@ func TestGetManagedClusters(t *testing.T) {
...
 		assert.Assert(t, count == 0)
 	})
 }
 
+func TestGetBridgeClusters(t *testing.T) {
+	ctx := context.Background()
+
+	t.Run("success", func(t *testing.T) {
+		fakeClient := setupFakeClientWithPGOScheme(t, true)
+		ctx, calls := setupLogCapture(ctx)
```
Comment (dsessler7) [inline internal/upgradecheck/header_test.go:528]: Would you mind adding go-doc-comment descriptions for these and other helper functions in `helpers_test.go`?

File: internal/upgradecheck/header_test.go [hunk 2]
```diff
@@ -554,3 +591,15 @@ func TestAddHeader(t *testing.T) {
...
+	t.Helper()
+	previous, ok := os.LookupEnv(key)
+	t.Cleanup(func() {
+		if ok {
+			os.Setenv(key, previous)
+		} else {
+			os.Unsetenv(key)
+		}
+	})
+}
```
Comment (dsessler7) [inline internal/upgradecheck/header_test.go:605]: Should this maybe go in `helpers_test.go`?

File: internal/upgradecheck/helpers_test.go [hunk 3]
```diff
@@ -87,7 +87,27 @@ func setupFakeClientWithPGOScheme(t *testing.T, includeCluster bool) crclient.Cl
...
+						Name: "elephant",
+					},
+				},
+			},
+		}
+
+		return fake.NewClientBuilder().
+			WithScheme(runtime.Scheme).
+			WithLists(pc).
+			WithLists(bcl).
```
Comment (dsessler7) [inline internal/upgradecheck/helpers_test.go:109]: :pick: These can be added in one `WithLists` call.

File: internal/upgradecheck/http_test.go [hunk 4]
```diff
@@ -72,7 +72,7 @@ func TestCheckForUpgrades(t *testing.T) {
 		}
 
 		res, header, err := checkForUpgrades(ctx, "", "4.7.3", backoff,
-			fakeClient, cfg, false)
+			fakeClient, cfg, false, "")
 		assert.NilError(t, err)
 		assert.Equal(t, res, `{"pgo_versions":[{"tag":"v5.0.4"},{"tag":"v5.0.3"},{"tag":"v5.0.2"},{"tag":"v5.0.1"},{"tag":"v5.0.0"}]}`)
 		checkData(t, header)
```
Comment (dsessler7) [inline internal/upgradecheck/http_test.go:78]: Should `checkData` be updated to check for the registration token? Should there be a case where we pass a non-empty string for the token and check that it makes it into the header properly?

**Review by dsessler7 — APPROVED** [review]
LGTM

File: internal/upgradecheck/header_test.go [hunk 5]
```diff
@@ -97,8 +107,11 @@ func TestGenerateHeader(t *testing.T) {
 	t.Run("success", func(t *testing.T) {
 		ctx, calls := setupLogCapture(ctx)
 
+		saveEnv(t, "PGO_FEATURE_GATES")
+		os.Setenv("PGO_FEATURE_GATES", "Tablespace=true")
```
[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/upgradecheck/header_test.go]: updated

File: internal/upgradecheck/helpers_test.go [hunk 6]
```diff
@@ -148,3 +178,17 @@ func setupNamespace(t testing.TB, cc crclient.Client) {
 	assert.NilError(t, cc.Create(ctx, ns))
 	t.Cleanup(func() { assert.Check(t, crclient.IgnoreNotFound(cc.Delete(ctx, ns))) })
 }
+
+// saveEnv preserves environment variables so that any modifications needed for
+// the tests can be undone once completed.
+func saveEnv(t testing.TB, key string) {
```
[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/upgradecheck/helpers_test.go]: Pulled that out of all the tests.

File: internal/upgradecheck/http_test.go [hunk 7]
```diff
@@ -47,10 +48,13 @@ func (m *MockClient) Do(req *http.Request) (*http.Response, error) {
 }
 
 func TestCheckForUpgrades(t *testing.T) {
-	fakeClient := setupFakeClientWithPGOScheme(t, false)
+	fakeClient := setupFakeClientWithPGOScheme(t, true)
 	ctx := logging.NewContext(context.Background(), logging.Discard())
 	cfg := &rest.Config{}
 
+	saveEnv(t, "PGO_FEATURE_GATES")
+	os.Setenv("PGO_FEATURE_GATES", "Tablespace=true")
```
[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/upgradecheck/http_test.go]: Updated

File: internal/upgradecheck/http.go [hunk 8]
```diff
@@ -125,24 +126,32 @@ type CheckForUpgradesScheduler struct {
...
+	OpenShift         bool
+	Refresh           time.Duration
+	URL, Version      string
+	RegistrationToken string
 }
 
 // ManagedScheduler creates a [CheckForUpgradesScheduler] and adds it to m.
-func ManagedScheduler(m manager.Manager, openshift bool, url, version string) error {
+func ManagedScheduler(m manager.Manager, openshift bool,
+	url, version string, registrationToken *jwt.Token) error {
```
[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/upgradecheck/http.go:137]: Hmmm, should the upgrade check mechanism instantiate a registration runner to make sure we get the current token if we ever change the restart behavior? Or I guess: leave the note now and change it in the future if we figure out we need it. (I'm leaving this comment here as a record of my slow thinking.)

[the remark this answers is not in the document]
Comment (benjaminjb) [reply internal/upgradecheck/http.go:142]: Added a comment, care to take a look.

**Review by dsessler7 — APPROVED** [review]

File: internal/upgradecheck/header.go [hunk 9]
```diff
@@ -33,24 +35,36 @@ var (
...
+	DeploymentID        string `json:"deployment_id"`
+	FeatureGatesEnabled string `json:"feature_gates_enabled"`
+	IsOpenShift         bool   `json:"is_open_shift"`
+	KubernetesEnv       string `json:"kubernetes_env"`
+	PGOClustersTotal    int    `json:"pgo_clusters_total"`
+	PGOVersion          string `json:"pgo_version"`
+	RegistrationToken   string `json:"registration_token"`
+	PGOInstaller        string `json:"pgo_installer"`
+	PGOInstallerOrigin  string `json:"pgo_installer_origin"`
+	BuildSource         string `json:"build_source"`
```
Comment (dsessler7) [inline internal/upgradecheck/header.go:48]: :pick: We should probably alphabetize these struct fields.

File: internal/upgradecheck/header.go [hunk 10]
```diff
@@ -33,24 +35,36 @@ var (
...
+		FeatureGatesEnabled: feature.ShowGates(ctx),
+		IsOpenShift:         isOpenShift,
+		DeploymentID:        ensureDeploymentID(ctx, crClient),
+		KubernetesEnv:       getServerVersion(ctx, cfg),
+		PGOClustersTotal:    getManagedClusters(ctx, crClient),
+		PGOVersion:          pgoVersion,
+		RegistrationToken:   registrationToken,
+		PGOInstaller:        os.Getenv("PGO_INSTALLER"),
+		PGOInstallerOrigin:  os.Getenv("PGO_INSTALLER_ORIGIN"),
+		BuildSource:         os.Getenv("BUILD_SOURCE"),
```
Comment (dsessler7) [inline internal/upgradecheck/header.go:67]: :pick: We should probably alphabetize these struct fields.

File: internal/upgradecheck/http.go [hunk 11]
```diff
@@ -125,24 +126,37 @@ type CheckForUpgradesScheduler struct {
 	Client crclient.Client
 	Config *rest.Config
 
-	OpenShift    bool
-	Refresh      time.Duration
-	URL, Version string
+	OpenShift         bool
+	Refresh           time.Duration
+	URL, Version      string
+	RegistrationToken string
```
Comment (dsessler7) [inline internal/upgradecheck/http.go:132]: :pick: We should probably alphabetize these.

File: internal/upgradecheck/http.go [hunk 12]
```diff
@@ -125,24 +126,37 @@ type CheckForUpgradesScheduler struct {
...
-		Refresh:   24 * time.Hour,
-		URL:       url,
-		Version:   version,
+		Client:            m.GetClient(),
+		Config:            m.GetConfig(),
+		OpenShift:         openshift,
+		Refresh:           24 * time.Hour,
+		URL:               url,
+		Version:           version,
+		RegistrationToken: token,
```
Comment (dsessler7) [inline internal/upgradecheck/http.go:159]: :pick: We should probably alphabetize these.

**Review by dsessler7 — APPROVED** [review]

**Review by tjmoore4 — APPROVED** [review]

**Pull request merged**